x/dclupgrade/handler_test.go

```diff
@@ -0,0 +1,432 @@
+// Copyright 2020 DSR Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//nolint:testpackage,lll, staticcheck
+package dclupgrade
+
+import (
+	"context"
+	"testing"
+
+	storetypes "github.com/cosmos/cosmos-sdk/store/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	"github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/require"
+	testconstants "github.com/zigbee-alliance/distributed-compliance-ledger/integration_tests/constants"
+	testkeeper "github.com/zigbee-alliance/distributed-compliance-ledger/testutil/keeper"
+	"github.com/zigbee-alliance/distributed-compliance-ledger/testutil/testdata"
+	dclauthtypes "github.com/zigbee-alliance/distributed-compliance-ledger/x/dclauth/types"
+	"github.com/zigbee-alliance/distributed-compliance-ledger/x/dclupgrade/keeper"
+	"github.com/zigbee-alliance/distributed-compliance-ledger/x/dclupgrade/types"
+)
+
+type TestSetup struct {
+	T *testing.T
+	// Cdc         *amino.Codec
+	Ctx           sdk.Context
+	Wctx          context.Context
+	Keeper        *keeper.Keeper
+	DclauthKeeper *DclauthKeeperMock
+	UpgradeKeeper *UpgradeKeeperMock
```

Comment:
It makes perfect sense to use Mocks, as we consider these tests as unit tests.
However, this module (unit) can be considered as an extension of the cosmos's upgrade module, so what if we use a real cosmos's UpgradeKeeper here? Will it simplify the tests?
For example
1) Instead of calling `AssertCalled` for `isContextWithCachedMultiStore`, it will be sufficient just to have a test where invalid Plan is proposed (if an expected error is returned, then we are sure that all needed methods have been called)
2) Instead of calling `AssertCalled` for `isContextWithNonCachedMultiStore`, we could just check that an entry in cosmos's upgrade keeper is created.

Reply:
As an option, tests of `x/dclupgrade` handler would be written in the described way. It might simplify the tests a bit, but would make them not pure unit tests.